### ISSUE TITLE
verbose mode does not show progress

### ISSUE THREAD

**Issue by Trass3r** [cmu-sei/pharos#25]
Usually you get a sense of progress out of the "processing function %u/%u" messages. But with -v3 or -v4 it only prints "Processing function %p"

Also there are no timestamps / time remaining according to the limits shown.

**Comment by sei-ccohen** [comment]
If your terminal is correctly configured, you should get progress bars that look like this:

PRT2[MARCH]: cfg:  26% [####-----------] 24674 bytes 2629 blks 73 funcs
OPTI[MARCH]: Function analysis:  26% [####-----------] 47

You should receive similar messages when higher levels of verbosity are on, but they're probably hard to "see" because they're automatically removed from the screen when a new message is emitted, which happens more frequently in higher verbosity levels.  The pretty progress bar is disabled when output is redirected to a file, because it's not clear how to "draw" it.  This could be improved, but it's met our debugging needs so far.

**Comment by Trass3r** [comment]
Indeed I used `tee` to save the output for later reference. Maybe it log the lower level stuff to a file directly.

**Comment by sei-ccohen** [comment]
If you're really interested in fine-tuning control of the log levels, the system is based on ROSE/Sawyer and is described in some detail here:

http://rosecompiler.org/ROSE_HTML_Reference/namespaceRose_1_1Diagnostics.html
http://rosecompiler.org/ROSE_HTML_Reference/namespaceSawyer_1_1Message.html

We have an option (--log) that will pass the [Facility control message](http://rosecompiler.org/ROSE_HTML_Reference/classSawyer_1_1Message_1_1Facilities.html#a5eee6e4f739534b2a2399142be6c5361) onto ROSE.  We've perhaps not been as diligent in assigning messages to useful log levels as we should be, but those links should make it easy for you to modify the system to get whatever logging you desire.  The progress bars are explained in there too.

**Comment by Trass3r** [comment]
I see, thanks. In the end I simply need enough realtime output to know if it's still doing something / making progress and maybe a more detailed log file for later.

**Comment by sei-ccohen** [comment]
It sounds like you eventually got what you needed for this issue, so I'm going to close it.

**Comment by Trass3r** [comment]
No not really. Even without tee I don't see any progress bar with -v3 or -v4.
I don't see how these --log options can be used to dump something to a file either.

**Comment by sei-ccohen** [comment]
I'm not certain I understand the problem exactly then.  I'm going to focus on this statement: "In the end I simply need enough realtime output to know if it's still doing something." 

I assume you're not in the first "partitioning" stage because ROSE is fairly good about issuing messages periodically during that step.  I'm also going to assume that you're not in in the second "program analysis" phase, because with --verbose enabled, it prints messages like "Processing function 0x..." before analyzing each function, which shows progress pretty steadily.   The progress bar does not show up well in interactive mode with --verbose on, but it is still there.   When redirected to a file, the progress bar is disabled (or altered at least).  But it should be easily to tell if progress is being made.  

Thus I conclude that (maybe?) you're in the Prolog analysis phase.  We have had problems with this phase never really completing for large input programs, so that seems like a reasonable guess.   Try adding --prolog-debug to enable debugging spew from Prolog (this is also on by default in the other script I pointed you to).  This should ensure that you're receiving regular debugging messages, although I won't promise that it actually fixes the underlying problem about why it's running so long. :-(  We've engaged the XSB authors on that question, and it seems to be a difficult question for them as well.

**Comment by sei-ccohen** [comment]
It's also possible that you were stuck in the partitioning stage, but that's a bug that we've fixed in this release compared to the previous one, so if you weren't receiving progress messages in the first (partitioning) stage, just try the new release.

**Comment by Trass3r** [comment]
No, partitioning is no problem and the serialization works nicely.

"Processing function 0x" does tell you where it is currently but not how much is left. Only if you look up the largest address, which is what I do now.

>  The progress bar does not show up well in interactive mode with --verbose on, but it is still there.

I didn't see any.

I did also have an issue with the Prolog phase lately. It made only negligible progress, burning full CPU in an emulate loop. Indeed there's no feedback there either.
It didn't really allocate much memory and there was some left. So it shouldn't be a memory issue. Though I'm not sure how much it needs and the program analysis memory wasn't released before starting the Prolog phase.

**Comment by Trass3r** [comment]
No, partitioning is no problem and the serialization works nicely.

"Processing function 0x" does tell you where it is currently but not how much is left. Also jumps around so you never know.

>  The progress bar does not show up well in interactive mode with --verbose on, but it is still there.

I didn't see any. Actually gone back to non-verbose mode to get it again. That combined with the `--timing`  option is crucial for knowing if the limits will work out. Would still be nice to have some verbose output to a file directly for later reference.

The Prolog phase is indeed hard to follow. The number of facts goes up and down which doesn't help building a feeling for how much time/effort is still required.

**Comment by sei-ccohen** [comment]
You are correct that it's very hard to tell how close to finished you are during the Prolog phase.  In part that because as humans we don't understand how to predict the final number of facts.  Because we can put a number on "done", the progress bar isn't very helpful.   This is obviously research software, and you're definitely digging into the problems that we consider "open problems" at this point.

Ideally the number of facts would more or less increase steadily during the Prolog phase.  It's normal for it it go up a few and then to dip a little before resuming it's upward trend.  When the facts go down, that's where we've made a bad "guess" and the Prolog backtracking has kicked in.  Unfortunately, we've seen behavior where the backtracking goes into an unproductive repetitive search pattern (e.g. ABCD^D^CD^D^BCD^D^CD^D...)  Where ^X is "not X".  It's a sign that we made a bad guess early on, and didn't catch it until much later.  That's not really supposed to happen, and probably means that our rules need improvement.  If we had guessed the correct answer instead of the incorrect answer, the entire problem could have been avoided.  It is often difficult to determine where we first went off track and why.  Alternatively, we could just detect the bad guess earlier.  Less ambiguous starting facts might also help reduce bad guesses.

But that "bug" isn't the most serious part of problem.  What's really harsh is that for some reason that we don't fully understand (and the XSB developers haven't really be able to explain either) each repetition of the pattern gets slower and slower.  Thus the run time is not just whatever that combinatorial expansion yields (which can be bad), it also has to include each successive step being slower.  You can see this when Prolog debugging is on and it begins to take a minute or more for each guess/retract cycle while the earlier guess/retract cycles even for the same guess previously took just seconds.  When we reach this point on an input program after 10-20 hours, we conclude that the program is "not likely to finish in any reasonable time", although we have had programs complete after more than 24 hours. :-(

We've found and fixed a bunch of problems that resulted in bad behavior like this, and there's some evidence that this problem is not a fundamental system design flaw, but more like a very difficult to fix implementation problem.  Of course it's also unclear how far we are from having fixed all of these cases, and what the upper bound on the input program size is before we encounter some other performance problem.

**Comment by sei-ccohen** [comment]
The October update including some new Prolog logging capabilities that should make it easier to tell how things are progressing.  They can be enabled with various combinations of --prolog-debug, --log="PLOG(all)" and --verbose=4.   We still have the fundamental problem that there's no way to know when we'll run out of guesses, so we still don't know how far we are from finishing.   I don't currently see anything that can be done about that.  If you still have concerns that should keep this issue open, please let us know (or close it).
